esp8266/esp8266.go

```diff
@@ -0,0 +1,245 @@
+//
+package esp8266
+
+import (
+	"machine"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Device wraps UART connection to the ESP8266.
+type Device struct {
+	bus machine.UART
+
+	// command responses that come back from the ESP8266
+	response []byte
+
+	// data received from a TCP/UDP connection forwarded by the ESP8266
+	socketdata []byte
+}
+
+// New returns a new esp8266-wifi driver. Pass in a fully configured UART bus.
+func New(b machine.UART) Device {
+	return Device{bus: b, response: make([]byte, 512), socketdata: make([]byte, 0, 1024)}
+}
+
+// Configure sets up the device for communication.
+func (d Device) Configure() {
+}
+
+// Connected checks if there is communication with the ESP8266.
+func (d *Device) Connected() bool {
+	d.Execute(Test)
+
+	// handle response here, should include "OK"
+	r := d.Response()
+	if strings.Contains(string(r), "OK") {
+		return true
+	}
+	return false
+}
+
+// Write raw bytes to the UART.
+func (d *Device) Write(b []byte) (n int, err error) {
+	return d.bus.Write(b)
+}
+
+// Read raw bytes from the UART.
+func (d *Device) Read(b []byte) (n int, err error) {
+	return d.bus.Read(b)
+}
+
+// how long in milliseconds to pause after sending AT commands
+const pause = 100
+
+// Execute sends an AT command to the ESP8266.
+func (d Device) Execute(cmd string) error {
+	_, err := d.Write([]byte("AT" + cmd + "\r\n"))
+	return err
+}
+
+// Query sends an AT command to the ESP8266 that returns the
+// current value for some configuration parameter.
+func (d Device) Query(cmd string) (string, error) {
+	_, err := d.Write([]byte("AT" + cmd + "?\r\n"))
+	return "", err
+}
+
+// Set sends an AT command with params to the ESP8266 for a
+// configuration value to be set.
+func (d Device) Set(cmd, params string) error {
+	_, err := d.Write([]byte("AT" + cmd + "=" + params + "\r\n"))
+	return err
+}
+
+// Version returns the ESP8266 firmware version info.
+func (d Device) Version() []byte {
+	d.Execute(Version)
+	return d.Response()
+}
+
+// Echo sets the ESP8266 echo setting.
+func (d Device) Echo(set bool) {
+	if set {
+		d.Execute(EchoConfigOn)
+	} else {
+		d.Execute(EchoConfigOff)
+	}
+	// TODO: check for success
+	d.Response()
+}
+
+// Reset restarts the ESP8266 firmware. Due to how the baud rate changes,
+// this messes up communication with the ESP8266 module. So make sure you know
+// what you are doing when you call this.
+func (d Device) Reset() {
+	d.Execute(Restart)
+	d.Response()
+}
+
+// ReadSocket returns the data that has already been read in from the responses.
+func (d *Device) ReadSocket(b []byte) (n int, err error) {
+	// make sure no data in buffer
+	d.Response()
+
+	count := len(b)
+	if len(b) >= len(d.socketdata) {
+		// copy it all, then clear socket data
+		count = len(d.socketdata)
+		copy(b, d.socketdata[:count])
+		d.socketdata = d.socketdata[:0]
+	} else {
+		// copy all we can, then keep the remaining socket data around
+		copy(b, d.socketdata[:count])
+		d.socketdata = d.socketdata[count:]
```

Comment:
Unfortunately, this is a memory leak. The buffer pointer moves forward every time this happens, and `append` just appends to the end.
It may not be perfect, but I think copying the data at the end to the start is the best thing to do here, unless you want to implement a circular buffer.

```go
// copy all we can, then keep the remaining socket data around
copy(b, d.socketdata[:count])
copy(d.socketdata, d.socketdata[count:])
d.socketdata = d.socketdata[:len(d.socketdata)-count]
```

Reply:
Change made, thanks.